File: frontend/src/types/workflow.ts
export type WorkflowStatus = "pending" | "running" | "completed" | "failed";
export type AgentStatus = "idle" | "running" | "completed" | "failed" | "skipped";

export interface AgentLog {
  id: string;
  agent_name: string;
  status: AgentStatus;
  error_message?: string;
  started_at?: string;
  completed_at?: string;
}

export interface PlannerLog {
  id: string;
  step_number: number;
  agent_selected: string;
  decision_reasoning: string;
  created_at: string;
}

export interface Workflow {
  id: string;
  status: WorkflowStatus;
  configuration_id: string;
  configuration_name?: string;
  campaign_name?: string;
  campaign_id?: string;
  started_at?: string;
  completed_at?: string;
  created_at: string;
  agent_logs?: AgentLog[];
}

export interface NodeState {
  agent_name: string;
  status: AgentStatus;
}

// ── Agent detail state (per-agent tracking for the viewer) ─────────

export interface AgentDetail {
  name: string;
  status: AgentStatus;
  duration?: number;
  outputSummary?: string;
  companyCount?: number;
  error?: string;
  substepName?: string;
  substepDetail?: string;
  progressPct?: number;
  index: number;
  total: number;
}

// ── Workflow-level stats ──────────────────────────────────────────

export interface WorkflowStats {
  companiesFound: number;
  validated: number;
  rejected: number;
  techAnalyzed: number;
  marketSignals: number;
  highUrgency: number;
  contactsFound: number;
  linkedinConfirmed: number;
  emailsFound: number;
  tierA: number;
  tierB: number;
  tierC: number;
  recommendations: number;
  avgEmailConfidence: number;
  avgCompositeScore: number;
}

// ── Activity feed entries ─────────────────────────────────────────

export type ActivityEventType =
  | "workflow_started"
  | "workflow_completed"
  | "workflow_failed"
  | "agent_started"
  | "agent_completed"
  | "agent_failed"
  | "agent_substep"
  | "company_discovered"
  | "company_validated"
  | "company_rejected"
  | "tech_detected"
  | "contact_found"
  | "qualification_scored"
  | "recommendation_created";

export interface ActivityEntry {
  id: string;
  timestamp: string;
  eventType: ActivityEventType;
  agentName: string;
  payload: Record<string, unknown>;
}

// ── WebSocket event envelope ──────────────────────────────────────

export interface WorkflowEvent {
  event_type: string;
  workflow_id: string;
  agent_name: string;
  data: Record<string, unknown>;
  timestamp: string;
}

// ── Campaign info ─────────────────────────────────────────────────

export interface CampaignInfo {
  name: string;
  strategySummary?: string;
  totalAgents?: number;
  activeAgents?: number;
  skippedAgents?: number;
  reasoning?: string;
}

// ── Discovery company (lightweight for activity list) ─────────────

export interface DiscoveredCompany {
  name: string;
  domain: string;
  sourceQuery?: string;
}

// ── Full results from /workflow/{id}/results ──────────────────────

export interface ContactResult {
  id: string;
  full_name: string;
  role: string;
  email?: string;
  phone?: string;
  linkedin_url?: string;
  confidence_score?: number;
  is_primary_persona: boolean;
}

export interface CompanyResult {
  id: string;
  name: string;
  domain: string;
  industry?: string;
  country?: string;
  city?: string;
  employee_count?: number;
  revenue_range?: string;
  funding_stage?: string;
  tech_stack_json?: string[];
  qualification_score?: number;
  score_breakdown?: Record<string, number>;
  status: string;
  rejection_reason?: string;
  logo_url?: string;
  description?: string;
  growth_rate?: number;
  contacts: ContactResult[];
  created_at?: string;
}

export interface WorkflowResultSummary {
  total_companies: number;
  validated: number;
  rejected: number;
  pending: number;
  with_scores: number;
  with_contacts: number;
  with_recommendations: number;
}

export interface WorkflowResult {
  summary: WorkflowResultSummary;
  companies: CompanyResult[];
  recommendations: any[];
}

// ── Agent card expandable detail ──────────────────────────────────

export interface AgentCardDetail {
  companies?: DiscoveredCompany[];
  validationBreakdown?: { check: string; passed: boolean; detail: string }[];
  topTechnologies?: string[];
  tierDistribution?: { A: number; B: number; C: number };
}

File: backend/schemas/workflow.py
"""
Pydantic schemas for workflow endpoints.
"""

from datetime import datetime
from typing import Optional

from pydantic import BaseModel


class AgentLogEntry(BaseModel):
    id: str
    agent_name: str
    status: str
    error_message: Optional[str] = None
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class WorkflowStartRequest(BaseModel):
    configuration_id: str
    override_icp_json: Optional[dict] = None


class WorkflowStatusResponse(BaseModel):
    id: str
    status: str
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    agent_logs: list[AgentLogEntry] = []
    campaign_name: Optional[str] = None

    class Config:
        from_attributes = True


class ContactSummary(BaseModel):
    id: str
    full_name: str
    role: str
    email: Optional[str] = None
    phone: Optional[str] = None
    linkedin_url: Optional[str] = None
    confidence_score: Optional[float] = None
    is_primary_persona: bool = False

    class Config:
        from_attributes = True


class CompanyDetail(BaseModel):
    id: str
    name: str
    domain: str
    industry: Optional[str] = None
    country: Optional[str] = None
    city: Optional[str] = None
    employee_count: Optional[int] = None
    revenue_range: Optional[str] = None
    funding_stage: Optional[str] = None
    tech_stack_json: Optional[list] = None
    qualification_score: Optional[float] = None
    score_breakdown_json: Optional[dict] = None
    score_breakdown: Optional[dict] = None
    status: str
    rejection_reason: Optional[str] = None
    logo_url: Optional[str] = None
    description: Optional[str] = None
    growth_rate: Optional[float] = None
    contacts: list[ContactSummary] = []
    created_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class RecommendationSummary(BaseModel):
    id: str
    company_id: str
    priority: str
    reason: str
    suggested_action: str
    confidence: float

    class Config:
        from_attributes = True


class WorkflowResultsSummary(BaseModel):
    total_companies: int = 0
    validated: int = 0
    rejected: int = 0
    pending: int = 0
    with_scores: int = 0
    with_contacts: int = 0
    with_recommendations: int = 0


class WorkflowResultsResponse(BaseModel):
    summary: WorkflowResultsSummary
    companies: list[CompanyDetail] = []
    recommendations: list[RecommendationSummary] = []


class RetryAgentRequest(BaseModel):
    agent_name: str


class WorkflowListItem(BaseModel):
    id: str
    status: str
    configuration_id: str
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    created_at: datetime

    class Config:
        from_attributes = True
